scalableminds/webknossos
# Slick: Replace Mongo with Postgresql

### TODO
- [x] create SQL DAOs for all collections, use them instead of mongo
- [x] progress reports
- [x] complex queries handle isDeleted – create views for all collections?
- [x] automatically create db on startup if it’s not present
- [x] update slick classes from schema only if changed
- [x] access definitions
- [x] CI / docker
- [x] add indexes
- [x] slick rejects token cleanup job?
- [x] tidy up code
- [x] update from-mongo migration script
- [x] test db data
- [x] use postgres test db in e2e tests
- [x] resolve: UNIQUE vs isActive?
- [x] how to do schema evolutions in the future? → update schema.sql + create "alter table" evolution files

### Follow-Up TODOs
- fix hot-reloading memory leak for local development
- After the Organizations PR (issue #935) is done, apply these changes here
- remove the now-unused mongo tools
- make the new code a lot DRYer

### Mailable description of changes:
WK now uses Relational Database Postgresql instead of Mongo

### Steps to test:
- pretty much the full api needs to be tested.
- special emphasis on access permissions (test with multiple users + teams) and deleted items, task management etc.
- test performance
- check console for any "SQL Error" logs
- not affected: tree operations / update actions
- test migration scripts in `tools/postgres/pg_migrate`

### Issues:
- fixes #2243 
- fixes #2159 
- fixes #2098

------
